File: TimberCruisingApp/ViewModels/CruiseDesignViewModel.swift
// Spec §3.1 REQ-PRJ-003/004. Configure cruise design (plot type, plot size
// or BAF, sampling scheme, grid spacing) and generate PlannedPlots via the
// Geo SamplingGenerator. Persists the design record and replaces all
// previously-generated PlannedPlots for this project.

import Foundation
import Models
import Persistence
import Geo

@MainActor
public final class CruiseDesignViewModel: ObservableObject {

    // MARK: - Editable form state

    @Published public var plotType: PlotType = .fixedArea
    @Published public var plotAreaAcresString: String = "0.1"
    @Published public var bafString: String = "20"
    @Published public var samplingScheme: SamplingScheme = .systematicGrid
    @Published public var gridSpacingMetersString: String = "150"
    @Published public var nPerStratumString: String = "10"
    @Published public var seedString: String = "1"

    // MARK: - Feedback

    @Published public private(set) var plannedCount: Int = 0
    @Published public private(set) var availableSpecies: [SpeciesConfig] = []
    @Published public private(set) var volumeEquationsById: [String: VolumeEquation] = [:]
    @Published public var errorMessage: String?
    @Published public var toastMessage: String?

    public let project: Project

    private var stratumRepository: (any StratumRepository)?
    private var plannedPlotRepository: (any PlannedPlotRepository)?
    private var designRepository: (any CruiseDesignRepository)?
    private var speciesRepository: (any SpeciesConfigRepository)?
    private var volumeEquationRepository: (any VolumeEquationRepository)?

    public init(project: Project) { self.project = project }

    public func configure(with environment: AppEnvironment) {
        if stratumRepository == nil { stratumRepository = environment.stratumRepository }
        if plannedPlotRepository == nil { plannedPlotRepository = environment.plannedPlotRepository }
        if designRepository == nil { designRepository = environment.cruiseDesignRepository }
        if speciesRepository == nil { speciesRepository = environment.speciesRepository }
        if volumeEquationRepository == nil { volumeEquationRepository = environment.volumeEquationRepository }
    }

    // MARK: - Validation

    public var isValid: Bool { validationMessage == nil }

    public var validationMessage: String? {
        if plotType == .fixedArea {
            guard let area = Double(plotAreaAcresString), area > 0 else {
                return "Plot area must be a positive number of acres."
            }
            _ = area
        } else {
            guard let baf = Double(bafString), baf > 0 else {
                return "BAF must be a positive number."
            }
            _ = baf
        }
        switch samplingScheme {
        case .systematicGrid:
            guard let spacing = Double(gridSpacingMetersString), spacing > 0 else {
                return "Grid spacing must be a positive number of metres."
            }
            _ = spacing
        case .stratifiedRandom:
            guard let n = Int(nPerStratumString), n > 0 else {
                return "Plots per stratum must be a positive integer."
            }
            _ = n
        case .manual:
            return nil
        }
        return nil
    }

    // MARK: - Actions

    public func generatePlannedPlots() {
        guard isValid else {
            errorMessage = validationMessage
            return
        }
        guard let stratumRepo = stratumRepository,
              let plotRepo = plannedPlotRepository,
              let designRepo = designRepository
        else { return }

        do {
            let strata = try stratumRepo.listByProject(project.id)
            guard !strata.isEmpty else {
                errorMessage = "Add at least one stratum before generating plots."
                return
            }

            let inputs: [SamplingGenerator.StratumInput] = try strata.map { s in
                let rings = try parseRings(from: s.polygonGeoJSON)
                return .init(stratumId: s.id, rings: rings)
            }

            let seed = UInt64(seedString) ?? 1
            let options = SamplingGenerator.GenerationOptions(
                projectId: project.id,
                scheme: samplingScheme,
                gridSpacingMeters: samplingScheme == .systematicGrid
                    ? Double(gridSpacingMetersString) : nil,
                nPerStratum: samplingScheme == .stratifiedRandom
                    ? Int(nPerStratumString) : nil,
                seed: seed
            )
            let plots = try SamplingGenerator.generate(strata: inputs, options: options)

            // Replace existing planned plots for this project.
            let existing = try plotRepo.listByProject(project.id)
            for p in existing { try plotRepo.delete(id: p.id) }
            for plot in plots { _ = try plotRepo.create(plot) }

            // Persist the design record (one per project).
            let existingDesigns = try designRepo.forProject(project.id)
            let design = CruiseDesign(
                id: existingDesigns.first?.id ?? UUID(),
                projectId: project.id,
                plotType: plotType,
                plotAreaAcres: plotType == .fixedArea
                    ? Float(plotAreaAcresString) : nil,
                baf: plotType == .variableRadius
                    ? Float(bafString) : nil,
                samplingScheme: samplingScheme,
                gridSpacingMeters: samplingScheme == .systematicGrid
                    ? Float(gridSpacingMetersString) : nil
            )
            if existingDesigns.first != nil {
                _ = try designRepo.update(design)
            } else {
                _ = try designRepo.create(design)
            }

            plannedCount = plots.count
            toastMessage = "Generated \(plots.count) planned plot\(plots.count == 1 ? "" : "s")."
        } catch {
            errorMessage = "Plot generation failed: \(error)"
        }
    }

    public func refresh() {
        guard let plotRepo = plannedPlotRepository,
              let designRepo = designRepository
        else { return }
        do {
            plannedCount = try plotRepo.listByProject(project.id).count
            if let existing = try designRepo.forProject(project.id).first {
                plotType = existing.plotType
                if let a = existing.plotAreaAcres { plotAreaAcresString = "\(a)" }
                if let b = existing.baf { bafString = "\(b)" }
                samplingScheme = existing.samplingScheme
                if let g = existing.gridSpacingMeters { gridSpacingMetersString = "\(g)" }
            }
            if let speciesRepo = speciesRepository {
                availableSpecies = try speciesRepo.list()
                    .sorted { $0.commonName < $1.commonName }
            }
            if let volRepo = volumeEquationRepository {
                let eqs = try volRepo.list()
                volumeEquationsById = Dictionary(
                    uniqueKeysWithValues: eqs.map { ($0.id, $0) })
            }
        } catch {
            errorMessage = "Failed to load cruise design: \(error)"
        }
    }

    // MARK: - GeoJSON → rings

    private func parseRings(from geojson: String) throws -> [[CoordinateConversions.LatLon]] {
        guard let data = geojson.data(using: .utf8),
              let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any]
        else {
            throw NSError(domain: "CruiseDesign", code: 1,
                          userInfo: [NSLocalizedDescriptionKey: "Invalid stored polygon JSON"])
        }
        guard obj["type"] as? String == "Polygon",
              let coords = obj["coordinates"] as? [[[Double]]]
        else {
            throw NSError(domain: "CruiseDesign", code: 2,
                          userInfo: [NSLocalizedDescriptionKey: "Stored geometry is not a Polygon"])
        }
        return coords.map { ring in
            ring.compactMap { pair in
                guard pair.count >= 2 else { return nil }
                return CoordinateConversions.LatLon(latitude: pair[1], longitude: pair[0])
            }
        }
    }
}

File: android/app/src/main/java/com/hcjeong/forestix/ui/screens/project/CruiseDesignViewModel.kt
// Port of iOS ViewModels/CruiseDesignViewModel.swift.
// Spec §3.1 REQ-PRJ-003/004. Configure cruise design (plot type, plot size
// or BAF, sampling scheme, grid spacing) and generate PlannedPlots via the
// Geo SamplingGenerator. Persists the design record and replaces all
// previously-generated PlannedPlots for this project.

package com.hcjeong.forestix.ui.screens.project

import com.hcjeong.forestix.AppEnvironment
import com.hcjeong.forestix.data.cruise.CruiseDesign
import com.hcjeong.forestix.data.cruise.CruiseDesignRepository
import com.hcjeong.forestix.data.cruise.PlannedPlotRepository
import com.hcjeong.forestix.data.cruise.PlotType
import com.hcjeong.forestix.data.cruise.Project
import com.hcjeong.forestix.data.cruise.SamplingScheme
import com.hcjeong.forestix.data.cruise.SpeciesConfig
import com.hcjeong.forestix.data.cruise.SpeciesConfigRepository
import com.hcjeong.forestix.data.cruise.StratumRepository
import com.hcjeong.forestix.data.cruise.VolumeEquation
import com.hcjeong.forestix.data.cruise.VolumeEquationRepository
import com.hcjeong.forestix.geo.CoordinateConversions
import com.hcjeong.forestix.geo.SamplingGenerator
import java.util.UUID
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import org.json.JSONObject

class CruiseDesignViewModel(val project: Project) {

    // MARK: - Editable form state

    val plotType = MutableStateFlow(PlotType.FIXED_AREA)
    val plotAreaAcresString = MutableStateFlow("0.1")
    val bafString = MutableStateFlow("20")
    val samplingScheme = MutableStateFlow(SamplingScheme.SYSTEMATIC_GRID)
    val gridSpacingMetersString = MutableStateFlow("150")
    val nPerStratumString = MutableStateFlow("10")
    val seedString = MutableStateFlow("1")

    // MARK: - Feedback

    private val _plannedCount = MutableStateFlow(0)
    val plannedCount: StateFlow<Int> = _plannedCount.asStateFlow()

    private val _availableSpecies = MutableStateFlow<List<SpeciesConfig>>(emptyList())
    val availableSpecies: StateFlow<List<SpeciesConfig>> = _availableSpecies.asStateFlow()

    private val _volumeEquationsById = MutableStateFlow<Map<String, VolumeEquation>>(emptyMap())
    val volumeEquationsById: StateFlow<Map<String, VolumeEquation>> =
        _volumeEquationsById.asStateFlow()

    val errorMessage = MutableStateFlow<String?>(null)
    val toastMessage = MutableStateFlow<String?>(null)

    private var stratumRepository: StratumRepository? = null
    private var plannedPlotRepository: PlannedPlotRepository? = null
    private var designRepository: CruiseDesignRepository? = null
    private var speciesRepository: SpeciesConfigRepository? = null
    private var volumeEquationRepository: VolumeEquationRepository? = null

    fun configure(environment: AppEnvironment) {
        if (stratumRepository == null) stratumRepository = environment.stratumRepository
        if (plannedPlotRepository == null) plannedPlotRepository = environment.plannedPlotRepository
        if (designRepository == null) designRepository = environment.cruiseDesignRepository
        if (speciesRepository == null) speciesRepository = environment.speciesConfigRepository
        if (volumeEquationRepository == null) {
            volumeEquationRepository = environment.volumeEquationRepository
        }
    }

    // MARK: - Validation

    val isValid: Boolean get() = validationMessage == null

    /// Recomputed from the current field values on every read; the screen
    /// re-reads it during recomposition so it tracks keystrokes like the
    /// iOS computed property.
    val validationMessage: String?
        get() {
            if (plotType.value == PlotType.FIXED_AREA) {
                val area = plotAreaAcresString.value.toDoubleOrNull()
                if (area == null || area <= 0) {
                    return "Plot area must be a positive number of acres."
                }
            } else {
                val baf = bafString.value.toDoubleOrNull()
                if (baf == null || baf <= 0) {
                    return "Basal area factor must be a positive number."
                }
            }
            when (samplingScheme.value) {
                SamplingScheme.SYSTEMATIC_GRID -> {
                    val spacing = gridSpacingMetersString.value.toDoubleOrNull()
                    if (spacing == null || spacing <= 0) {
                        return "Grid spacing must be a positive number of metres."
                    }
                }
                SamplingScheme.STRATIFIED_RANDOM -> {
                    val n = nPerStratumString.value.toIntOrNull()
                    if (n == null || n <= 0) {
                        return "Plots per stratum must be a positive integer."
                    }
                }
                SamplingScheme.MANUAL -> return null
            }
            return null
        }

    // MARK: - Actions

    suspend fun generatePlannedPlots() {
        if (!isValid) {
            errorMessage.value = validationMessage
            return
        }
        val stratumRepo = stratumRepository ?: return
        val plotRepo = plannedPlotRepository ?: return
        val designRepo = designRepository ?: return

        try {
            val strata = stratumRepo.listByProject(project.id)
            if (strata.isEmpty()) {
                errorMessage.value = "Add at least one stratum before generating plots."
                return
            }

            val inputs = strata.map { s ->
                SamplingGenerator.StratumInput(
                    stratumId = s.id,
                    rings = parseRings(s.polygonGeoJSON),
                )
            }

            val seed = seedString.value.toULongOrNull() ?: 1uL
            val options = SamplingGenerator.GenerationOptions(
                projectId = project.id,
                scheme = samplingScheme.value,
                gridSpacingMeters = if (samplingScheme.value == SamplingScheme.SYSTEMATIC_GRID) {
                    gridSpacingMetersString.value.toDoubleOrNull()
                } else null,
                nPerStratum = if (samplingScheme.value == SamplingScheme.STRATIFIED_RANDOM) {
                    nPerStratumString.value.toIntOrNull()
                } else null,
                seed = seed,
            )
            val plots = SamplingGenerator.generate(strata = inputs, options = options)

            // Replace existing planned plots for this project.
            val existing = plotRepo.listByProject(project.id)
            for (p in existing) plotRepo.delete(p.id)
            for (plot in plots) plotRepo.create(plot)

            // Persist the design record (one per project).
            val existingDesigns = designRepo.forProject(project.id)
            val design = CruiseDesign(
                id = existingDesigns.firstOrNull()?.id ?: UUID.randomUUID(),
                projectId = project.id,
                plotType = plotType.value,
                plotAreaAcres = if (plotType.value == PlotType.FIXED_AREA) {
                    plotAreaAcresString.value.toFloatOrNull()
                } else null,
                baf = if (plotType.value == PlotType.VARIABLE_RADIUS) {
                    bafString.value.toFloatOrNull()
                } else null,
                samplingScheme = samplingScheme.value,
                gridSpacingMeters = if (samplingScheme.value == SamplingScheme.SYSTEMATIC_GRID) {
                    gridSpacingMetersString.value.toFloatOrNull()
                } else null,
            )
            if (existingDesigns.firstOrNull() != null) {
                designRepo.update(design)
            } else {
                designRepo.create(design)
            }

            _plannedCount.value = plots.size
            toastMessage.value =
                "Generated ${plots.size} planned plot${if (plots.size == 1) "" else "s"}."
        } catch (e: Exception) {
            errorMessage.value = "Plot generation failed: $e"
        }
    }

    suspend fun refresh() {
        val plotRepo = plannedPlotRepository ?: return
        val designRepo = designRepository ?: return
        try {
            _plannedCount.value = plotRepo.listByProject(project.id).size
            val existing = designRepo.forProject(project.id).firstOrNull()
            if (existing != null) {
                plotType.value = existing.plotType
                existing.plotAreaAcres?.let { plotAreaAcresString.value = it.toString() }
                existing.baf?.let { bafString.value = it.toString() }
                samplingScheme.value = existing.samplingScheme
                existing.gridSpacingMeters?.let { gridSpacingMetersString.value = it.toString() }
            }
            speciesRepository?.let { repo ->
                _availableSpecies.value = repo.list().sortedBy { it.commonName }
            }
            volumeEquationRepository?.let { repo ->
                _volumeEquationsById.value = repo.list().associateBy { it.id }
            }
        } catch (e: Exception) {
            errorMessage.value = "Failed to load cruise design: $e"
        }
    }

    // MARK: - GeoJSON → rings

    private fun parseRings(geojson: String): List<List<CoordinateConversions.LatLon>> {
        val obj = try {
            JSONObject(geojson)
        } catch (_: Exception) {
            throw IllegalArgumentException("Invalid stored polygon JSON")
        }
        if (obj.optString("type") != "Polygon") {
            throw IllegalArgumentException("Stored geometry is not a Polygon")
        }
        val coords = obj.optJSONArray("coordinates")
            ?: throw IllegalArgumentException("Stored geometry is not a Polygon")
        val rings = mutableListOf<List<CoordinateConversions.LatLon>>()
        for (ri in 0 until coords.length()) {
            val ringArray = coords.getJSONArray(ri)
            val ring = mutableListOf<CoordinateConversions.LatLon>()
            for (pi in 0 until ringArray.length()) {
                val pair = ringArray.getJSONArray(pi)
                if (pair.length() >= 2) {
                    ring.add(
                        CoordinateConversions.LatLon(
                            latitude = pair.getDouble(1),
                            longitude = pair.getDouble(0),
                        )
                    )
                }
            }
            rings.add(ring)
        }
        return rings
    }
}
